atc/worker/gardenruntime/worker.go

```diff
@@ -713,3 +640,89 @@ func (worker *Worker) GardenClient() gclient.Client {
 func (worker *Worker) BaggageclaimClient() baggageclaim.Client {
 	return worker.bcClient
 }
+
+func (worker *Worker) acquireVolumeStreamingLock(
+	ctx context.Context,
+	volume string,
+	destination string,
+) (lock.Lock, bool, error) {
+	// if caching of streamed volumes is not enabled then we should not globally
+	// lock the volume streaming since steps cannot share the streamed volumes as caches
+	if !atc.EnableCacheStreamedVolumes {
+		return lock.NoopLock{}, true, nil
+	}
+
+	logger := lagerctx.FromContext(ctx)
+	return worker.db.LockFactory.Acquire(logger, lock.NewVolumeStreamingLockID(volume, destination))
+}
+
+func (worker *Worker) findOrStreamVolume(
+	ctx context.Context,
+	privileged bool,
+	teamID int,
+	container db.CreatingContainer,
+	artifact runtime.Artifact,
+	mountPath string,
+	delegate runtime.BuildStepDelegate,
+) (Volume, error) {
+	logger := lagerctx.FromContext(ctx)
+	var tick *time.Ticker
+
+	_, inputPath := path.Split(mountPath)
+	if inputPath == "" {
+		inputPath = "for image"
+	}
+
+	for {
+		localVolume, ok, err := worker.findVolumeForArtifact(ctx, teamID, artifact)
+		if err != nil {
+			logger.Error("failed-to-find-volume-for-artifact", err)
+			return Volume{}, err
+		}
+		if ok && localVolume.DBVolume().WorkerName() == worker.Name() {
+			volume := localVolume.(Volume)
+			return volume, nil
+		}
+
+		lock, acquired, err := worker.acquireVolumeStreamingLock(ctx, artifact.Handle(), container.WorkerName())
```

Comment:
I think we should use resource cache id for the lock.

Say, worker-1 and worker-2 both have an image, and now 2 builds on worker-3 need the image. The two builds' nested `image-get` may return different artifacts from worker-1 and worker-2, but they actually point to the same image. In this case, if we use volume handle for lock, worker-3 will still duplicate streaming the image.

Reply:
Yeah that's a nice extra optimization. It means casting from `runtime.Artifact` to `runtime.Volume` again which feels pretty bad (all artifacts currently are volumes) but we can make it work easily enough. I know you bumped into similar issues with the original cache-streamed-volumes work.